File: shopify/app/app/lib/proofmark/cookies.ts
/**
 * Server-side cookie helpers (storefront cookies set via the App Proxy
 * response). Same `showad_*` naming as @showad/nextjs-sdk.
 *
 * NOTE: Shopify's app proxy returns the response under the merchant's
 * storefront domain, so any `Set-Cookie` here lands on the merchant origin
 * and is sent back on subsequent storefront requests.
 */

export const COOKIE_PREFIX = 'showad';
export const COOKIE_FINGERPRINT = `${COOKIE_PREFIX}_fingerprint`;
export const COOKIE_TOKEN = `${COOKIE_PREFIX}_token`;
export const COOKIE_CREATOR = `${COOKIE_PREFIX}_creator`;
export const COOKIE_TICKET = `${COOKIE_PREFIX}_ticket`;
export const COOKIE_VERIFIED = `${COOKIE_PREFIX}_verified`;
export const COOKIE_EXPIRES = `${COOKIE_PREFIX}_expires`;

export interface CookieOptions {
  path?: string;
  maxAge?: number;
  domain?: string;
  httpOnly?: boolean;
  secure?: boolean;
  sameSite?: 'Strict' | 'Lax' | 'None';
}

export function parseCookieHeader(header: string | null | undefined): Record<string, string> {
  const out: Record<string, string> = {};
  if (!header) return out;
  for (const part of header.split(';')) {
    const idx = part.indexOf('=');
    if (idx < 0) continue;
    const name = part.slice(0, idx).trim();
    const value = part.slice(idx + 1).trim();
    if (!name) continue;
    try {
      out[name] = decodeURIComponent(value);
    } catch {
      out[name] = value;
    }
  }
  return out;
}

export function serializeCookie(name: string, value: string, options: CookieOptions = {}): string {
  if (!/^[!#$%&'*+\-.0-9A-Z^_`a-z|~]+$/.test(name)) {
    throw new Error('Invalid cookie name');
  }
  const parts = [`${name}=${encodeURIComponent(value)}`];
  if (options.maxAge !== undefined) parts.push(`Max-Age=${Math.floor(options.maxAge)}`);
  parts.push(`Path=${options.path || '/'}`);
  if (options.domain) parts.push(`Domain=${options.domain}`);
  if (options.httpOnly) parts.push('HttpOnly');
  if (options.secure) parts.push('Secure');
  parts.push(`SameSite=${options.sameSite || 'Lax'}`);
  return parts.join('; ');
}

export interface VerificationCookieData {
  token: string;
  creatorHash: string;
  ticketId?: string | null;
  fingerprint?: string | null;
  expiresAt?: number | null;
  cookieMaxAge: number;
  secure: boolean;
}

export function buildVerificationCookies(data: VerificationCookieData): string[] {
  const base: CookieOptions = {
    path: '/',
    maxAge: data.cookieMaxAge,
    secure: data.secure,
    sameSite: 'Lax',
  };
  const cookies = [
    serializeCookie(COOKIE_TOKEN, data.token, { ...base, httpOnly: true }),
    serializeCookie(COOKIE_VERIFIED, '1', base),
    serializeCookie(COOKIE_CREATOR, data.creatorHash, base),
  ];
  if (data.ticketId) cookies.push(serializeCookie(COOKIE_TICKET, data.ticketId, base));
  if (data.fingerprint) cookies.push(serializeCookie(COOKIE_FINGERPRINT, data.fingerprint, base));
  if (typeof data.expiresAt === 'number') {
    cookies.push(serializeCookie(COOKIE_EXPIRES, String(data.expiresAt), base));
  }
  return cookies;
}

export function buildClearCookies(secure: boolean): string[] {
  const opts: CookieOptions = { path: '/', maxAge: 0, secure, sameSite: 'Lax' };
  return [
    serializeCookie(COOKIE_TOKEN, '', { ...opts, httpOnly: true }),
    serializeCookie(COOKIE_VERIFIED, '', opts),
    serializeCookie(COOKIE_CREATOR, '', opts),
    serializeCookie(COOKIE_TICKET, '', opts),
    serializeCookie(COOKIE_EXPIRES, '', opts),
  ];
}

File: python/src/proofmark_showad/cookies.py
"""Cookie name constants and helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Optional

from .config import ShowAdConfig

COOKIE_FINGERPRINT = "fingerprint"
COOKIE_TOKEN = "token"
COOKIE_CREATOR = "creator"
COOKIE_TICKET = "ticket"
COOKIE_VERIFIED = "verified"
COOKIE_EXPIRES = "expires"
COOKIE_META = "meta"

ALL_COOKIE_SUFFIXES: tuple = (
    COOKIE_FINGERPRINT,
    COOKIE_TOKEN,
    COOKIE_CREATOR,
    COOKIE_TICKET,
    COOKIE_VERIFIED,
    COOKIE_EXPIRES,
)


def cookie_name(prefix: str, suffix: str) -> str:
    return f"{prefix}_{suffix}"


def all_cookie_names(prefix: str) -> Iterable[str]:
    for suffix in ALL_COOKIE_SUFFIXES:
        yield cookie_name(prefix, suffix)


@dataclass
class CookieSpec:
    """Description of a cookie to set or clear in a framework-agnostic way."""

    name: str
    value: str
    max_age: int
    http_only: bool = False
    secure: bool = False
    same_site: str = "lax"
    path: str = "/"

    @property
    def is_clear(self) -> bool:
        return self.max_age <= 0 and self.value == ""


def build_set_cookie_specs(
    config: ShowAdConfig,
    *,
    token: Optional[str],
    creator_hash: Optional[str],
    ticket_id: Optional[str],
    token_expiry: Optional[int],
    secure: bool,
) -> list:
    """Build a list of :class:`CookieSpec` to set verification cookies."""

    max_age = config.cookie_max_age
    same_site = config.cookie_same_site
    prefix = config.cookie_prefix

    specs: list = []
    if token:
        specs.append(
            CookieSpec(
                name=cookie_name(prefix, COOKIE_TOKEN),
                value=token,
                max_age=max_age,
                http_only=True,
                secure=secure,
                same_site=same_site,
            )
        )
        specs.append(
            CookieSpec(
                name=cookie_name(prefix, COOKIE_VERIFIED),
                value="1",
                max_age=max_age,
                secure=secure,
                same_site=same_site,
            )
        )
    if creator_hash:
        specs.append(
            CookieSpec(
                name=cookie_name(prefix, COOKIE_CREATOR),
                value=creator_hash,
                max_age=max_age,
                secure=secure,
                same_site=same_site,
            )
        )
    if ticket_id:
        specs.append(
            CookieSpec(
                name=cookie_name(prefix, COOKIE_TICKET),
                value=ticket_id,
                max_age=max_age,
                secure=secure,
                same_site=same_site,
            )
        )
    if token_expiry is not None:
        specs.append(
            CookieSpec(
                name=cookie_name(prefix, COOKIE_EXPIRES),
                value=str(token_expiry),
                max_age=max_age,
                secure=secure,
                same_site=same_site,
            )
        )
    return specs


def build_clear_cookie_specs(config: ShowAdConfig, *, secure: bool) -> list:
    prefix = config.cookie_prefix
    same_site = config.cookie_same_site
    return [
        CookieSpec(
            name=cookie_name(prefix, suffix),
            value="",
            max_age=0,
            http_only=(suffix == COOKIE_TOKEN),
            secure=secure,
            same_site=same_site,
        )
        for suffix in (
            COOKIE_TOKEN,
            COOKIE_VERIFIED,
            COOKIE_CREATOR,
            COOKIE_TICKET,
            COOKIE_EXPIRES,
        )
    ]


__all__ = [
    "COOKIE_FINGERPRINT",
    "COOKIE_TOKEN",
    "COOKIE_CREATOR",
    "COOKIE_TICKET",
    "COOKIE_VERIFIED",
    "COOKIE_EXPIRES",
    "COOKIE_META",
    "ALL_COOKIE_SUFFIXES",
    "CookieSpec",
    "cookie_name",
    "all_cookie_names",
    "build_set_cookie_specs",
    "build_clear_cookie_specs",
]
